R/match_spec.R

```diff
@@ -67,106 +48,103 @@
 #' \code{\link{load_lib}()} loads the Open Specy reference library into an \R
 #' object of choice
 #'
-#' @importFrom rlang .data
-#' @importFrom stats approx cor
-#' @importFrom dplyr %>% inner_join mutate group_by ungroup summarize select arrange desc top_n
-#'
 #' @export
-match_spec <- function(x, ...) {
-  UseMethod("match_spec")
+correlate_spectra <- function(object, ...) {
+    UseMethod("correlate_spectra")
 }
 
-#' @rdname match_spec
-#'
 #' @export
-match_spec.formula <- function(formula, data = NULL, ...) {
-  if (missing(formula) || (length(formula) != 3L) ||
-      (length(attr(terms(formula[-2L]), "term.labels")) != 1L))
-    stop("'formula' missing or incorrect")
+correlate_spectra.default <- function(object, ...) {
+    stop("object needs to be of class 'OpenSpecy'")
+}
 
-  mf <- model.frame(formula, data)
-  lst <- as.list(mf)
-  names(lst) <- c("y", "x")
+#' @export
+correlate_spectra.OpenSpecy <- function(object, library, na.rm = T, ...){
+    if(sum(object$wavenumber %in% library$wavenumber) < 3){
+        stop("There are less than 3 matching wavenumbers in the objects you are trying to correlate, this won't work for correlation analysis. Consider first conforming the spectra to the same wavenumbers.")
+    }
+    cor(library$spectra[library$wavenumber %in% object$wavenumber,][,lapply(.SD, make_rel, na.rm = na.rm)][,lapply(.SD, mean_replace)],
+        object$spectra[object$wavenumber %in% library$wavenumber,][,lapply(.SD, make_rel, na.rm = na.rm)][,lapply(.SD, mean_replace)],
+        ...)
+}
 
-  do.call("match_spec", c(lst, list(...)))
+#' @export
+identify_spectra <- function(cor_matrix, object, library, top_n = NULL,
+                             add_library_metadata = NULL,
+                             add_object_metadata = NULL, ...){
+    if(is.numeric(top_n) && top_n > ncol(library$spectra)){
+        top_n = NULL
+        message("top_n was larger than the number of spectra in the library, returning all matches")
+    }
+
+    data.table(object_id = colnames(object$spectra),
+                                   library_id = rep(colnames(library$spectra),
+                                               each = ncol(object$spectra)),
+                                   match_val = c(cor_matrix)) %>%
+        { if (is.numeric(top_n)) .[order(-match_val), head(.SD, top_n), by = object_id] else .} %>%
+        { if (is.character(add_library_metadata)) left_join(., library$metadata, by = c("library_id" = add_library_metadata)) else . } %>%
+        { if (is.character(add_object_metadata)) left_join(., object$metadata, by = c("object_id" = add_object_metadata)) else . }
```

Comment:
The `left_join()` commands here are the only ones left requiring the dplyr package. I'd like to convert them to data.table commands but the function gives me an error once I add the `add_library_metadata` or `add_object_metadata` arguments. How's the function supposed to work?

Reply:
The add_library_metadata options is supposed to be a character value of the name of the column in the library object that needs to be joined to the library id column. And the add_object_metadata is the same but for the object metadata to the object_id

Reply:
Done. Tests still pass so I hope I didn't break it.